src/coreclr/jit/gentree.cpp

```diff
@@ -30261,85 +30391,83 @@ GenTree* Compiler::gtFoldExprHWIntrinsic(GenTreeHWIntrinsic* tree)
         }
     }
 
+#if defined(FEATURE_MASKED_HW_INTRINSICS)
     if (tree->OperIsConvertMaskToVector())
     {
         GenTree* op = op1;
 
-        if (!op->OperIsHWIntrinsic())
-        {
-            return tree;
-        }
-
-        unsigned            simdBaseTypeSize = genTypeSize(simdBaseType);
-        GenTreeHWIntrinsic* cvtOp            = op->AsHWIntrinsic();
-
-        if (!cvtOp->OperIsConvertVectorToMask())
+        if (op->OperIsHWIntrinsic())
         {
-            return tree;
-        }
+            unsigned            simdBaseTypeSize = genTypeSize(simdBaseType);
+            GenTreeHWIntrinsic* cvtOp            = op->AsHWIntrinsic();
 
-        if ((genTypeSize(cvtOp->GetSimdBaseType()) != simdBaseTypeSize))
-        {
-            // We need the operand to be the same kind of mask; otherwise
-            // the bitwise operation can differ in how it performs
-            return tree;
-        }
+            if (cvtOp->OperIsConvertVectorToMask())
+            {
+                if ((genTypeSize(cvtOp->GetSimdBaseType()) == simdBaseTypeSize))
+                {
+                    // We need the operand to be the same kind of mask; otherwise
+                    // the bitwise operation can differ in how it performs
 
 #if defined(TARGET_XARCH)
-        GenTree* vectorNode = cvtOp->Op(1);
+                    GenTree* vectorNode = cvtOp->Op(1);
 #elif defined(TARGET_ARM64)
-        GenTree* vectorNode = cvtOp->Op(2);
+                    GenTree* vectorNode = cvtOp->Op(2);
 #else
 #error Unsupported platform
 #endif // !TARGET_XARCH && !TARGET_ARM64
 
-        DEBUG_DESTROY_NODE(op, tree);
-        INDEBUG(vectorNode->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);
+                    DEBUG_DESTROY_NODE(op, tree);
+                    INDEBUG(vectorNode->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);
 
-        return vectorNode;
+                    return vectorNode;
+                }
+            }
+        }
     }
 
     if (tree->OperIsConvertVectorToMask())
     {
-        GenTree* op = op1;
+        GenTree* op        = op1;
+        bool     tryHandle = false;
 
-#if defined(TARGET_ARM64)
-        if (!op->OperIsHWIntrinsic(NI_Sve_CreateTrueMaskAll))
+#if defined(TARGET_XARCH)
+        tryHandle = op->OperIsHWIntrinsic();
+#elif defined(TARGET_ARM64)
+        if (op->OperIsHWIntrinsic() && op->OperIsHWIntrinsic(NI_Sve_CreateTrueMaskAll))
```

Comment:
Should this condition be checking `op2->OperIsHWIntrinsic()` instead of `op->OperIsHWIntrinsic`? Right now, it is possible for us to try to use `op2` as a `GT_HWINTRINSIC` node when it isn't one. The SVE tests are currently failing for me with the assertion `OperIs(GT_HWINTRINSIC)`, because of the line `GenTreeHWIntrinsic* cvtOp = op->AsHWIntrinsic();` below.

Reply:
OperIsHWIntrinsic is always safe, it simply checks if the node type is GT_HWINTRINSIC

It looks like there might be a missing nested check that op2 is itself a hwintrinsic after we’ve confirmed the root op is createtruemaskall

Really we should be normalizing CreateTrueMaskAll as CNS_MSK instead and simply not having it as part of CvtVectorToMask (it’s an implementation detail of codegen that has no real impact on HIR), but that was a more in depth change and so wasn’t done in this pr

Reply:
fixed in https://github.com/dotnet/runtime/pull/104998